{{cookiecutter.project_slug}}/.gitlab-ci.yml

```diff
@@ -0,0 +1,131 @@
+# templates
+# ------------
+
+.deploy-template:
+  image: alpine:latest
+  script:
+    - echo "Deploying to ${CI_ENVIRONMENT_NAME}"
+  variables:
+    GIT_STRATEGY: none  # prevents fetching repo (we only need the artifact here)
+
+# -------------
+# Pipeline Definition
+
+variables:
+  PIP_CACHE_DIR: $CI_PROJECT_DIR/.cache/pip
+  {%- if cookiecutter.package_manager == 'conda' %}
+  CONDA_PKGS_DIRS: $CI_PROJECT_DIR/.cache/conda
+  {%- elif cookiecutter.package_manager == 'poetry' %}
+  POETRY_VERSION: 1.0.5
+  POETRY_CACHE_DIR: $CI_PROJECT_DIR/.cache/poetry
+  {%- endif %}
+
+cache: &python_cache
+  paths:
+    - $PIP_CACHE_DIR
+    {%- if cookiecutter.package_manager == 'conda' %}
+    - $CONDA_PKGS_DIRS/*.tar.bz2
+    - $CONDA_PKGS_DIRS/urls.txt
+    {%- elif cookiecutter.package_manager == 'poetry' %}
+    - $POETRY_CACHE_DIR
+    {%- endif %}
+
+stages:
+  - build-wheel
+  - test-unit
+  - deploy-nonprod
+  - deploy-prod
+
+# Build and Test
+{% if cookiecutter.package_manager == 'poetry' -%}
+build-wheel:
+  image: python:3.7-stretch
+  stage: build-wheel
+  artifacts:
+    name: 'app-wheel'
+    paths:
+      - dist/{{ cookiecutter.module_name }}-*.whl
+    expire_in: 1 week
+  before_script:
+    - pip install poetry==$POETRY_VERSION
+  script:
+    - poetry build -f wheel
+{% else -%}
+build-wheel:
+  image: python:3.7-alpine
+  stage: build-wheel
+  artifacts:
+    name: 'app-wheel'
+    paths:
+      - dist/{{ cookiecutter.module_name }}-*.whl
+    expire_in: 1 week
+  script:
+    - python setup.py dist
+{%- endif %}
+
+{% if cookiecutter.package_manager == 'poetry' -%}
+test-unit:
+  stage: test-unit
+  image: python:3.7-stretch
+  cache:
+    <<: *python_cache
```

Comment:
I'm not sure how this cache statement works. why is it necessary here?

Reply:
This is a yaml functionality, called aliases and anchors. Basically I defined the anchor on top of the script and can reuse it with the alias in several places of the script. See e.g. here: http://blogs.perl.org/users/tinita/2019/05/reusing-data-with-yaml-anchors-aliases-and-merge-keys.html

Reply:
ok, but why do we have to define the cache again in this job? Isn't the global cache declaration being used for every job?

Reply:
Yes, I actually just added this to show how this could be reused and extended within the jobs but I didn't actually extend it so I think this is probably more confusing than helpful. I will change it to have it only as a global cache setting.